File: backend/src/main/java/pl/com/ww/mesh/atlas/integration/application/service/IntegrationPipelineService.java
package pl.com.ww.mesh.atlas.integration.application.service;

import jakarta.persistence.criteria.Predicate;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pl.com.ww.mesh.atlas.global.domain.exception.AtlasException;
import pl.com.ww.mesh.atlas.integration.application.dto.IntegrationPipelineCreateRequest;
import pl.com.ww.mesh.atlas.integration.application.dto.IntegrationPipelineDto;
import pl.com.ww.mesh.atlas.integration.application.dto.IntegrationPipelineSummaryDto;
import pl.com.ww.mesh.atlas.integration.application.dto.IntegrationPipelineUpdateRequest;
import pl.com.ww.mesh.atlas.integration.application.mapper.IntegrationPipelineMapper;
import pl.com.ww.mesh.atlas.integration.domain.exception.IntegrationDatasourceNotFoundException;
import pl.com.ww.mesh.atlas.integration.domain.exception.IntegrationPipelineNotFoundException;
import pl.com.ww.mesh.atlas.integration.domain.model.IntegrationDatasourceEntity;
import pl.com.ww.mesh.atlas.integration.domain.model.IntegrationPipelineEntity;
import pl.com.ww.mesh.atlas.integration.domain.model.PipelineStatus;
import pl.com.ww.mesh.atlas.integration.domain.model.TargetEntityType;
import pl.com.ww.mesh.atlas.integration.infrastructure.camel.CamelDslValidatorService;
import pl.com.ww.mesh.atlas.integration.infrastructure.persistence.IntegrationDatasourceRepository;
import pl.com.ww.mesh.atlas.integration.infrastructure.persistence.IntegrationPipelineRepository;
import pl.com.ww.mesh.atlas.integration.infrastructure.scheduler.PipelineSchedulerService;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class IntegrationPipelineService {

    private final IntegrationPipelineRepository pipelineRepository;
    private final IntegrationDatasourceRepository datasourceRepository;
    private final IntegrationPipelineMapper mapper;
    private final CamelDslValidatorService camelDslValidator;
    private final PipelineSchedulerService schedulerService;

    public Page<IntegrationPipelineSummaryDto> findAll(
            Boolean active, String code, String name,
            PipelineStatus status, TargetEntityType targetEntity,
            Pageable pageable) {
        return pipelineRepository.findAll(buildSpec(active, code, name, status, targetEntity), pageable)
                .map(mapper::mapSummary);
    }

    private Specification<IntegrationPipelineEntity> buildSpec(
            Boolean active, String code, String name,
            PipelineStatus status, TargetEntityType targetEntity) {
        return (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();
            if (active != null) {
                predicates.add(cb.equal(root.get("active"), active));
            }
            if (code != null && !code.isBlank()) {
                predicates.add(cb.like(cb.lower(root.get("code")), "%" + code.toLowerCase() + "%"));
            }
            if (name != null && !name.isBlank()) {
                predicates.add(cb.like(cb.lower(root.get("name")), "%" + name.toLowerCase() + "%"));
            }
            if (status != null) {
                predicates.add(cb.equal(root.get("status"), status));
            }
            if (targetEntity != null) {
                predicates.add(cb.equal(root.get("targetEntity"), targetEntity));
            }
            return cb.and(predicates.toArray(new Predicate[0]));
        };
    }

    public IntegrationPipelineDto findById(UUID id) {
        return mapper.map(load(id));
    }

    public String getDsl(UUID id) {
        IntegrationPipelineEntity entity = load(id);
        if (entity.getCamelXmlDsl() == null) {
            throw new AtlasException("No DSL uploaded for pipeline: " + id,
                    "DSL_NOT_FOUND", id.toString(), HttpStatus.NOT_FOUND);
        }
        return entity.getCamelXmlDsl();
    }

    @Transactional
    public IntegrationPipelineDto create(IntegrationPipelineCreateRequest request) {
        if (pipelineRepository.existsByCode(request.code())) {
            throw new AtlasException("Pipeline code already exists: " + request.code(),
                    "DUPLICATE_CODE", request.code(), HttpStatus.CONFLICT);
        }
        IntegrationDatasourceEntity datasource = datasourceRepository.findById(request.datasourceId())
                .orElseThrow(() -> new IntegrationDatasourceNotFoundException(request.datasourceId()));

        IntegrationPipelineEntity entity = mapper.map(request);
        entity.setDatasource(datasource);
        entity = pipelineRepository.save(entity);

        Instant nextAt = schedulerService.scheduleOrUpdate(entity);
        entity.setNextExecutionAt(nextAt);

        return mapper.map(entity);
    }

    @Transactional
    public IntegrationPipelineDto update(UUID id, IntegrationPipelineUpdateRequest request) {
        IntegrationPipelineEntity entity = load(id);
        IntegrationDatasourceEntity datasource = datasourceRepository.findById(request.datasourceId())
                .orElseThrow(() -> new IntegrationDatasourceNotFoundException(request.datasourceId()));
        mapper.updateEntity(request, entity);
        entity.setDatasource(datasource);
        entity.setStatus(request.status());
        entity = pipelineRepository.save(entity);

        Instant nextAt = schedulerService.scheduleOrUpdate(entity);
        entity.setNextExecutionAt(nextAt);

        return mapper.map(entity);
    }

    @Transactional
    public IntegrationPipelineDto uploadDsl(UUID id, String xmlDsl) {
        camelDslValidator.validate(xmlDsl);
        IntegrationPipelineEntity entity = load(id);
        entity.setCamelXmlDsl(xmlDsl);
        return mapper.map(pipelineRepository.save(entity));
    }

    @Transactional
    public void deactivate(UUID id) {
        IntegrationPipelineEntity entity = load(id);
        entity.setActive(false);
        entity.setNextExecutionAt(null);
        pipelineRepository.save(entity);
        schedulerService.removeJobIfExists(id);
    }

    IntegrationPipelineEntity load(UUID id) {
        return pipelineRepository.findById(id)
                .orElseThrow(() -> new IntegrationPipelineNotFoundException(id));
    }
}

File: frontend/src/app/integration/service/integration-pipeline.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@environments/environment';
import {
  IntegrationPipelineCreateRequest,
  IntegrationPipelineDto,
  IntegrationPipelineSummaryDto,
  IntegrationPipelineUpdateRequest,
  Page,
  SyncTriggerDto,
} from '../model/integration.model';

@Injectable({ providedIn: 'root' })
export class IntegrationPipelineService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.APIUrl}/api/v1/integration/pipelines`;

  findAll(
    page = 0,
    size = 20,
    filters: {
      active?: boolean | null;
      code?: string | null;
      name?: string | null;
      status?: string | null;
      targetEntity?: string | null;
    } = {},
  ): Observable<Page<IntegrationPipelineSummaryDto>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (filters.active != null) params = params.set('active', String(filters.active));
    if (filters.code?.trim()) params = params.set('code', filters.code.trim());
    if (filters.name?.trim()) params = params.set('name', filters.name.trim());
    if (filters.status) params = params.set('status', filters.status);
    if (filters.targetEntity) params = params.set('targetEntity', filters.targetEntity);
    return this.http.get<Page<IntegrationPipelineSummaryDto>>(this.baseUrl, { params });
  }

  findById(id: string): Observable<IntegrationPipelineDto> {
    return this.http.get<IntegrationPipelineDto>(`${this.baseUrl}/${id}`);
  }

  create(request: IntegrationPipelineCreateRequest): Observable<IntegrationPipelineDto> {
    return this.http.post<IntegrationPipelineDto>(this.baseUrl, request);
  }

  update(id: string, request: IntegrationPipelineUpdateRequest): Observable<IntegrationPipelineDto> {
    return this.http.put<IntegrationPipelineDto>(`${this.baseUrl}/${id}`, request);
  }

  deactivate(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  uploadDsl(id: string, xmlContent: string): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${id}/dsl`, xmlContent, {
      headers: { 'Content-Type': 'text/plain' },
    });
  }

  getDslUrl(id: string): string {
    return `${this.baseUrl}/${id}/dsl`;
  }

  triggerSync(id: string): Observable<SyncTriggerDto> {
    return this.http.post<SyncTriggerDto>(`${this.baseUrl}/${id}/sync`, {});
  }
}
